### DISCUSSION TITLE
Dynamically adding/removing instruments to a meter at runtime.

### DISCUSSION THREAD

**Discussion by Simonl9l** [open-telemetry/opentelemetry-dotnet-instrumentation#3893] [Q&A] [answered]
Hi, trying to switch over to Open Telemetry from `System.Metrics.Diagnostics`, to using the DI service builder, and `AddOpenTelemtry()` etc.

Is there a way to dynamically add or remove instruments, at runtime? How about same for Metrics?

My application can hot reload configuration that determined what metrics and instruments are created, so not sure how to do this if its not know at startup.

Thanks!

**Comment by RassK** [comment]
@Simonl9l OTel SDK officially does not support this. You may do similar what we do in OTel Auto for "lazy instrumentation". Start from here https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs

**Reply by Simonl9l** [reply]
Hi @RassK - thanks for the reply (applogies it's taken while to get back to you).

On initial inspection of the LazyLoader, and trying to track back though the Repo, I'm not finding how this ties back to the `AddOpenTeletmetry()` and the builder as part of startup code. It's all bit of a fire hose.

A few more hints would be hugely appreciated is you could perhaps provide a small example to get me going!

**Reply by RassK** [reply, marked as answer]
It's a standalone part. No connection to `AddOpenTelemetry()` but `AddSource()`. 

`LazyInstrumentationLoader` is just stateholder of Initializers (for example `HttpClientInitializer`). Initializer specifies when an specific instrumentation instance is created and when invoked, then creates the instance.

For eg `HttpClientInitializer` is waiting for `System.Net.Http.dll` to load. If loaded it creates instance of `OpenTelemetry.Instrumentation.Http.HttpClientInstrumentation`.

`EnvironmentConfigurationTracerHelper.Wrappers.AddHttpClientInstrumentation()` wires up all the parts. First adds a lazy initializer to listen for assembly load events. Second always subscribes to System.Diagnostic.DiagnosticSource sources (since you cannot subscribe after ProviderBuilder is built).
